health/observer.go

```diff
@@ -0,0 +1,80 @@
+package health
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+)
+
+type Observer struct {
+	checks               []Check
+	checkLock            sync.RWMutex
+	runInterval          time.Duration
+	runLock              sync.RWMutex
+	runErr               error
+	initialChecksRunOnce sync.Once // is used to ensure collectInitialChecks is run exactly once by a consuming entity
+}
+
+func NewObserver(interval time.Duration) *Observer {
+	return &Observer{
+		checks:      []Check{},
+		runInterval: interval,
+	}
+}
+
+func (c *Observer) AddChecks(checks ...Check) {
+	c.checkLock.Lock()
+	c.checks = append(c.checks, checks...)
+	c.checkLock.Unlock()
+}
+
+func (c *Observer) Status() error {
```

Comment:
Another possibility here would be to have a `CheckResult` object with `Name` and `Error`, which would be `nil` for successful checks, so a caller could report all checks in their output.

Reply:
We could have the `CheckResult` look something like this:
```
type CheckResult struct {
    Name string
    Error error
}

func (c CheckResult) String() {
    status := "ok"
    if c.Error != nil {
        status = c.Error.String()
    }
    return fmt.Sprintf("%s: %s", c.Name, status)
}
```